File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        retArr = []
        subset = []

        def dfs(i):
            if i >= len(nums):
                retArr.append(subset.copy())
                return
            subset.append(nums[i])
            print(subset)
            dfs(i+1)
            subset.pop()
            dfs(i+1)

        dfs(0)
        return retArr

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> retArr;

    void dfs(vector<int>& candidates, vector<int>& subset, int target, int index) {
        if (target == 0) {
            retArr.push_back(subset);
            return;
        }
        if (index >= candidates.size() or target < 0) {
            return;
        }

        subset.push_back(candidates[index]);
        dfs(candidates, subset, target - candidates[index], index + 1);
        subset.pop_back();

        while ((index + 1) < candidates.size() and candidates[index] == candidates[index + 1]) {
            index++;
        }
        dfs(candidates, subset, target, index + 1);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> subset;
        dfs(candidates, subset, target, 0);
        return retArr;
    }
};
